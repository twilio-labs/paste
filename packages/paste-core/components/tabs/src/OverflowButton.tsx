import { Box, BoxProps, BoxStyleProps } from "@twilio-paste/box";
import { ChevronLeftIcon } from "@twilio-paste/icons/esm/ChevronLeftIcon";
import { ChevronRightIcon } from "@twilio-paste/icons/esm/ChevronRightIcon";
import { useTheme } from "@twilio-paste/theme";
import React from "react";

import { TabsContext } from "./TabsContext";

interface OverflowButtonProps {
  onClick: () => void;
  position: "left" | "right";
  visible?: boolean;
  element?: BoxProps["element"];
}

const Styles: BoxStyleProps = {
  color: "colorTextIcon",
  _hover: {
    color: "colorTextPrimaryStrongest",
    cursor: "pointer",
  },
};

const InverseStyles: BoxStyleProps = {
  color: "colorTextIconInverse",
  _hover: {
    color: "colorTextInverseWeak",
    cursor: "pointer",
  },
};

export const OverflowButton: React.FC<OverflowButtonProps> = ({ onClick, position, visible, element = "TAB_LIST" }) => {
  const theme = useTheme();
  const { variant } = React.useContext(TabsContext);
  const isInverse = variant?.includes("inverse");
  const Chevron = position === "left" ? ChevronLeftIcon : ChevronRightIcon;
  if (!visible) return null;

  return (
    <Box
      onClick={onClick}
      aria-hidden={true}
      display="flex"
      alignItems="center"
      justifyContent="center"
      width="sizeSquare70"
      boxShadow={isInverse ? theme.shadows.shadowScrollInverse : theme.shadows.shadowScroll}
      element={`${element}_OVERFLOW_BUTTON_${position.toUpperCase()}`}
      {...(isInverse ? InverseStyles : Styles)}
    >
      <Chevron decorative={true} />
    </Box>
  );
};

OverflowButton.displayName = "OverflowButton";
