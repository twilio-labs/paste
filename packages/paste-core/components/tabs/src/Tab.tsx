import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, BoxStyleProps, safelySpreadBoxProps} from '@twilio-paste/box';
import {TabPrimitive} from '@twilio-paste/tabs-primitive';
import {TabsContext} from './TabsContext';
import {Orientation, Variants} from './types';

// TODO:
// Split vertical tabs into a separate component
// because fitted tabs do nothing when orientation
// is vertical, which feels broken. Vertical tabs
// shouldn't get the fitted variant.
const getTabBoxStyles = (orientation: Orientation, variant: Variants): BoxStyleProps => {
  switch (orientation) {
    case 'vertical':
      return {
        borderLeftColor: 'transparent',
        borderLeftStyle: 'solid',
        borderLeftWidth: 'borderWidth20',
        color: 'colorTextWeak',
        display: 'block',
        marginBottom: 'space40',
        paddingBottom: 'space30',
        paddingLeft: 'space50',
        paddingRight: 'space50',
        paddingTop: 'space30',
        _last: {
          marginBottom: 'space0',
        },
        _selected: {
          borderLeftColor: 'colorBorderPrimary',
          color: 'colorTextLink',
        },
        _hover: {
          borderLeftColor: 'colorBorderPrimaryDarker',
          color: 'colorTextLinkDarker',
        },
        _disabled: {
          borderLeftColor: 'transparent',
          color: 'colorTextWeaker',
        },
        _focus: {
          borderLeftColor: 'colorBorderPrimaryDarker',
          boxShadow: 'shadowFocus',
          color: 'colorTextLinkDarker',
          outline: 'none',
        },
      };
    case 'horizontal':
    default:
      return {
        borderBottomColor: 'transparent',
        borderBottomStyle: 'solid',
        borderBottomWidth: 'borderWidth20',
        color: 'colorTextWeak',
        display: 'inline-block',
        flexBasis: variant === 'fitted' ? '50%' : undefined,
        flexGrow: variant === 'fitted' ? 1 : undefined,
        flexShrink: variant === 'fitted' ? 1 : undefined,
        marginRight: variant === 'fitted' ? 'space0' : 'space70',
        minWidth: 'sizeSquare130',
        paddingBottom: 'space40',
        paddingLeft: 'space20',
        paddingRight: 'space20',
        paddingTop: 'space40',
        textAlign: variant === 'fitted' ? 'center' : undefined,
        _last: {
          marginRight: 'space0',
        },
        _selected: {
          borderBottomColor: 'colorBorderPrimary',
          color: 'colorTextLink',
        },
        _hover: {
          borderBottomColor: 'colorBorderPrimaryDarker',
          color: 'colorTextLinkDarker',
        },
        _disabled: {
          borderBottomColor: 'transparent',
          color: 'colorTextWeaker',
        },
        _focus: {
          borderBottomColor: 'colorBorderPrimaryDarker',
          boxShadow: 'shadowFocus',
          color: 'colorTextLinkDarker',
          outline: 'none',
        },
      };
  }
};

export interface TabProps {
  id?: string | undefined;
  focusable?: boolean | undefined;
  disabled?: boolean | undefined;
  children: React.ReactNode;
}

const Tab = React.forwardRef<HTMLDivElement, TabProps>(({children, ...tabProps}, ref) => {
  const tab = React.useContext(TabsContext);
  const boxStyles = React.useMemo(() => getTabBoxStyles(tab.orientation, tab.variant), [tab.orientation, tab.variant]);
  return (
    <TabPrimitive {...(tab as any)} {...tabProps} ref={ref}>
      {(props: TabProps) => {
        return (
          <Box
            {...safelySpreadBoxProps(props)}
            {...boxStyles}
            as="span"
            cursor={props['aria-disabled'] ? 'not-allowed' : 'pointer'}
            fontSize="fontSize30"
            fontWeight="fontWeightSemibold"
            overflow={tab.orientation !== 'vertical' ? 'hidden' : undefined}
            position="relative"
            textOverflow={tab.orientation !== 'vertical' ? 'ellipsis' : undefined}
            transition="border-color 100ms ease, color 100ms ease"
            whiteSpace={tab.orientation !== 'vertical' ? 'nowrap' : undefined}
          >
            {children}
          </Box>
        );
      }}
    </TabPrimitive>
  );
});

if (process.env.NODE_ENV === 'development') {
  Tab.propTypes = {
    id: PropTypes.string,
    focusable: PropTypes.bool,
    disabled: PropTypes.bool,
  };
}

Tab.displayName = 'Tab';

export {Tab};
