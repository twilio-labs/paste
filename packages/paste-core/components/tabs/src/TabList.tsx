import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import {TabPrimitiveList} from '@twilio-paste/tabs-primitive';
import {TabsContext} from './TabsContext';
import {Variants} from './types';

export interface TabListProps {
  'aria-label': string;
  disabled?: boolean | undefined;
  focusable?: boolean | undefined;
  children: React.ReactNode;
  variant?: Variants;
}

const HorizontalTabList: React.FC = ({children}) => (
  <Box
    display="flex"
    borderBottomWidth="borderWidth10"
    borderBottomColor="colorBorderWeak"
    borderBottomStyle="solid"
    marginBottom="space60"
  >
    {children}
  </Box>
);

const VerticalTabList: React.FC = ({children}) => (
  <Box
    borderLeftWidth="borderWidth10"
    borderLeftColor="colorBorderWeak"
    borderLeftStyle="solid"
    marginRight="space110"
    minWidth="size20"
    maxWidth="size40"
  >
    {children}
  </Box>
);

const TabList = React.forwardRef<HTMLDivElement, TabListProps>(({children, variant, ...props}, ref) => {
  const tab = React.useContext(TabsContext);
  const TabListWrapper = tab.orientation === 'vertical' ? VerticalTabList : HorizontalTabList;
  return (
    <TabPrimitiveList {...(tab as any)} {...props} ref={ref}>
      <TabListWrapper>{children}</TabListWrapper>
    </TabPrimitiveList>
  );
});

if (process.env.NODE_ENV === 'development') {
  TabList.propTypes = {
    'aria-label': PropTypes.string.isRequired,
    focusable: PropTypes.bool,
    disabled: PropTypes.bool,
  };
}

TabList.displayName = 'TabList';
export {TabList};
