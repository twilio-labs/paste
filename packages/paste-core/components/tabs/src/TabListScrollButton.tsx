import * as React from "react";
import { Box } from "@twilio-paste/box";
import { Button } from "@twilio-paste/button";
import { ChevronLeftIcon } from "@twilio-paste/icons/esm/ChevronLeftIcon";
import { ChevronRightIcon } from "@twilio-paste/icons/esm/ChevronRightIcon";

export const TabListScrollButton: React.FC<{ direction: "left" | "right"; label: string }> = ({ direction, label }) => {
  return (
    <Box
      position="fixed"
      height="38px"
      zIndex="zIndex10"
      backgroundColor="colorBackgroundBody"
      display="flex"
      alignItems="center"
      right={direction === "right" ? "0" : "auto"} // currently 0 of the page, make 0 of the tab list
    >
      <Button
        display="grid"
        alignItems="center"
        justifyContent="center"
        variant="reset"
        width="28px"
        height="28px"
        borderStyle="solid"
        borderColor="colorBorderWeaker"
        borderWidth="borderWidth10"
        borderRadius="borderRadius20"
        backgroundColor="colorBackgroundBody"
        boxShadow="shadowLow"
      >
        {direction === "left" && (
          <ChevronLeftIcon
            decorative={false}
            title={label}
            size="sizeIcon20"
            color="colorTextWeak"
          />
        )}
        {direction === "right" && (
          <ChevronRightIcon
            decorative={false}
            title={label}
            size="sizeIcon20"
            color="colorTextWeak"
          />
        )}
      </Button>
    </Box>
  );
};

TabListScrollButton.displayName = "TabListScrollButton";
