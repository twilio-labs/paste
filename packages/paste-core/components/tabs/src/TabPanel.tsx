import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box} from '@twilio-paste/box';
import type {BoxStyleProps} from '@twilio-paste/box';
import {TabPrimitivePanel} from '@twilio-paste/tabs-primitive';
import {TabsContext} from './TabsContext';

export const tabPanelStyles = {
  borderRadius: 'borderRadius20',
  _focusVisible: {
    boxShadow: 'shadowFocus',
    outline: 'none',
  },
} as Partial<BoxStyleProps>;

export interface TabPanelProps {
  id?: string | undefined;
  tabId?: string | undefined;
  children: React.ReactNode;
}

const TabPanel = React.forwardRef<HTMLDivElement, TabPanelProps>(({children, ...props}, ref) => {
  const tab = React.useContext(TabsContext);
  return (
    <TabPrimitivePanel {...(tab as any)} {...tabPanelStyles} {...props} as={Box} ref={ref}>
      {children}
    </TabPrimitivePanel>
  );
});

if (process.env.NODE_ENV === 'development') {
  TabPanel.propTypes = {
    id: PropTypes.string,
    tabId: PropTypes.string,
  };
}

TabPanel.displayName = 'TabPanel';
export {TabPanel};
