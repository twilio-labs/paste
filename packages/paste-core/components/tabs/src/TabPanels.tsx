import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';

import {TabsContext} from './TabsContext';
import {getElementName} from './utils';

export interface TabPanelsProps {
  children: React.ReactNode;
  element?: BoxProps['element'];
}

const TabPanels = React.forwardRef<HTMLDivElement, TabPanelsProps>(({children, element, ...props}, ref) => {
  const {orientation} = React.useContext(TabsContext);
  const elementName = getElementName(orientation, 'TAB_PANELS', element);
  return (
    <Box {...safelySpreadBoxProps(props)} element={elementName} width="100%" ref={ref}>
      {children}
    </Box>
  );
});

if (process.env.NODE_ENV === 'development') {
  TabPanels.propTypes = {
    element: PropTypes.string,
  };
}

TabPanels.displayName = 'TabPanels';
export {TabPanels};
