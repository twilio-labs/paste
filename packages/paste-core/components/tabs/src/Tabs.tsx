import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Flex} from '@twilio-paste/flex';
import {useTabPrimitiveState, TabPrimitiveInitialState, TabPrimitiveStateReturn} from '@twilio-paste/tabs-primitive';
import {TabsContext} from './TabsContext';
import {Variants} from './types';

export interface TabStateReturn extends TabPrimitiveStateReturn {
  [key: string]: any;
}

export interface TabsProps extends TabPrimitiveInitialState {
  state?: TabStateReturn;
  variant?: Variants;
}

// Set orientation to horizontal because undefined enables all arrow key movement
const Tabs = React.forwardRef<HTMLDivElement, TabsProps>(
  ({children, orientation = 'horizontal', state, variant, ...initialState}, ref) => {
    const tab = state || useTabPrimitiveState({orientation, ...initialState});
    const value = React.useMemo(() => ({...tab, variant}), [...Object.values(tab), variant]);
    const returnValue = <TabsContext.Provider value={value}>{children}</TabsContext.Provider>;

    if (tab.orientation === 'vertical') {
      return (
        <Flex ref={ref} wrap={false} vAlignContent="stretch">
          {returnValue}
        </Flex>
      );
    }
    return returnValue;
  }
);

if (process.env.NODE_ENV === 'development') {
  Tabs.propTypes = {
    selectedId: PropTypes.string,
    orientation: PropTypes.oneOf(['horizontal', 'vertical', null]),
    variant: PropTypes.oneOf(['fitted', null]),
  };
}

Tabs.displayName = 'Tabs';
export {Tabs};

export {useTabPrimitiveState as useTabState};
