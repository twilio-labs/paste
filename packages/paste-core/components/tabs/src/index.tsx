export * from './Tabs';
export * from './Tab';
export * from './TabList';
export * from './TabPanel';
export * from './TabPanels';
