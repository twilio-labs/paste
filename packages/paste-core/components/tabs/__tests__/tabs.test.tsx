import * as React from 'react';
import {render, screen} from '@testing-library/react';
import axe from '../../../../../.jest/axe-helper';
import {HorizontalTabsExample, VerticalTabsExample, StateHookExample} from '../stories/index.stories';
import {Tabs, Tab, TabList, TabPanels, TabPanel} from '../src';

describe('Tabs', () => {
  describe('Render', () => {
    it('horizontally and vertically', () => {
      const {asFragment} = render(
        <>
          <HorizontalTabsExample />
          <VerticalTabsExample />
        </>
      );
      expect(asFragment()).toMatchSnapshot();
    });

    it('relevant html and aria attributes', () => {
      const [tabOneId, tabTwoId, tabThreeId, panelOneId, panelTwoId, panelThreeId] = [...new Array(6)].map(
        (_, i) => `${i}`
      );

      const ManualIdExample: React.FC = () => {
        return (
          <Tabs orientation="horizontal" selectedId={tabOneId} baseId="">
            <TabList aria-label="My tabs">
              <Tab id={tabOneId}>Tab 1 is a long tab name because the server sent a long tab name</Tab>
              <Tab id={tabTwoId}>Tab 2</Tab>
              <Tab id={tabThreeId}>Tab 3</Tab>
            </TabList>
            <TabPanels>
              <TabPanel tabId={tabOneId} id={panelOneId}>
                Tab 1
              </TabPanel>
              <TabPanel tabId={tabTwoId} id={panelTwoId}>
                Tab 1
              </TabPanel>
              <TabPanel tabId={tabThreeId} id={panelThreeId}>
                Tab 1
              </TabPanel>
            </TabPanels>
          </Tabs>
        );
      };

      render(<ManualIdExample />);

      const renderedTabList = screen.getByRole('tablist');
      expect(renderedTabList.getAttribute('aria-orientation')).toEqual('horizontal');
      expect(renderedTabList.getAttribute('aria-label')).toEqual('My tabs');

      const [TabOne, TabTwo, TabThree] = screen.queryAllByRole('tab');

      expect(TabOne.getAttribute('aria-controls')).toBe(panelOneId);
      expect(TabOne.getAttribute('aria-selected')).toBe('true');
      expect(TabOne.getAttribute('id')).toBe(tabOneId);
      expect(TabOne.getAttribute('tabindex')).toBe('0');

      expect(TabTwo.getAttribute('aria-controls')).toBe(panelTwoId);
      expect(TabTwo.getAttribute('aria-selected')).toBe('false');
      expect(TabTwo.getAttribute('id')).toBe(tabTwoId);
      expect(TabTwo.getAttribute('tabindex')).toBe('-1');

      expect(TabThree.getAttribute('aria-controls')).toBe(panelThreeId);
      expect(TabThree.getAttribute('aria-selected')).toBe('false');
      expect(TabThree.getAttribute('id')).toBe(tabThreeId);
      expect(TabThree.getAttribute('tabindex')).toBe('-1');

      let activePanel = screen.queryByRole('tabpanel');
      expect(activePanel.getAttribute('aria-labelledby')).toBe(tabOneId);
      expect(activePanel.getAttribute('id')).toBe(panelOneId);
      expect(activePanel.getAttribute('tabindex')).toBe('0');

      TabTwo.click();
      activePanel = screen.queryByRole('tabpanel');
      expect(activePanel.getAttribute('aria-labelledby')).toBe(tabTwoId);
      expect(activePanel.getAttribute('id')).toBe(panelTwoId);
      expect(activePanel.getAttribute('tabindex')).toBe('0');
    });

    it('should render tabs using the state prop and go to the next tab on button click', () => {
      render(<StateHookExample />);

      const [ButtonOne] = screen.queryAllByRole('button');

      let activePanel = screen.queryByRole('tabpanel');
      expect(activePanel.getAttribute('aria-labelledby')).toBe('state-hook-tab-example-1');
      expect(activePanel.getAttribute('id')).toBe('state-hook-tab-example-3');
      expect(activePanel.getAttribute('tabindex')).toBe('0');

      ButtonOne.click();
      activePanel = screen.queryByRole('tabpanel');
      expect(activePanel.getAttribute('aria-labelledby')).toBe('state-hook-tab-example-2');
      expect(activePanel.getAttribute('id')).toBe('state-hook-tab-example-4');
      expect(activePanel.getAttribute('tabindex')).toBe('0');
    });
  });

  describe('Accessibility', () => {
    it('Should have no accessibility violations', async () => {
      const {container} = render(<HorizontalTabsExample />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
