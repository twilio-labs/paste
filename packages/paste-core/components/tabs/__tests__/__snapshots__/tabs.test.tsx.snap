// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Render horizontally and vertically 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  margin-bottom: space70;
  font-size: fontSize30;
  line-height: lineHeight40;
  font-weight: fontWeightNormal;
  color: colorText;
}

<p
    class="emotion-0"
    color="colorText"
    font-size="fontSize30"
    font-weight="fontWeightNormal"
  >
    These tabs have full keyboard support. Try tabbing and using arrow keys.
  </p>
  .emotion-3 {
  box-sizing: border-box;
  display: flex;
  border-bottom-width: borderWidth10;
  border-bottom-color: colorBorderWeak;
  border-bottom-style: solid;
  margin-bottom: space60;
}

.emotion-0 {
  box-sizing: border-box;
  border-bottom-color: transparent;
  border-bottom-style: solid;
  border-bottom-width: borderWidth20;
  color: colorTextWeak;
  display: inline-block;
  margin-right: space70;
  min-width: sizeSquare130;
  padding-bottom: space40;
  padding-left: space20;
  padding-right: space20;
  padding-top: space40;
  cursor: pointer;
  font-size: fontSize30;
  font-weight: fontWeightSemibold;
  overflow: hidden;
  position: relative;
  text-overflow: ellipsis;
  -webkit-transition: border-color 100ms ease,color 100ms ease;
  transition: border-color 100ms ease,color 100ms ease;
  white-space: nowrap;
}

.emotion-0:last-of-type {
  margin-right: space0;
}

.emotion-0[aria-selected=true] {
  border-bottom-color: colorBorderPrimary;
  color: colorTextLink;
}

.emotion-0:hover {
  border-bottom-color: colorBorderPrimaryStronger;
  color: colorTextLinkStronger;
}

.emotion-0:disabled,
.emotion-0:disabled:focus,
.emotion-0:disabled:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover {
  border-bottom-color: transparent;
  color: colorTextWeaker;
}

.emotion-0:focus {
  border-bottom-color: colorBorderPrimaryStronger;
  box-shadow: shadowFocus;
  color: colorTextLinkStronger;
  outline: none;
}

.emotion-1 {
  box-sizing: border-box;
  border-bottom-color: transparent;
  border-bottom-style: solid;
  border-bottom-width: borderWidth20;
  color: colorTextWeak;
  display: inline-block;
  margin-right: space70;
  min-width: sizeSquare130;
  padding-bottom: space40;
  padding-left: space20;
  padding-right: space20;
  padding-top: space40;
  cursor: not-allowed;
  font-size: fontSize30;
  font-weight: fontWeightSemibold;
  overflow: hidden;
  position: relative;
  text-overflow: ellipsis;
  -webkit-transition: border-color 100ms ease,color 100ms ease;
  transition: border-color 100ms ease,color 100ms ease;
  white-space: nowrap;
}

.emotion-1:last-of-type {
  margin-right: space0;
}

.emotion-1[aria-selected=true] {
  border-bottom-color: colorBorderPrimary;
  color: colorTextLink;
}

.emotion-1:hover {
  border-bottom-color: colorBorderPrimaryStronger;
  color: colorTextLinkStronger;
}

.emotion-1:disabled,
.emotion-1:disabled:focus,
.emotion-1:disabled:hover,
.emotion-1[aria-disabled=true],
.emotion-1[aria-disabled=true]:focus,
.emotion-1[aria-disabled=true]:hover {
  border-bottom-color: transparent;
  color: colorTextWeaker;
}

.emotion-1:focus {
  border-bottom-color: colorBorderPrimaryStronger;
  box-shadow: shadowFocus;
  color: colorTextLinkStronger;
  outline: none;
}

<div
    aria-label="My tabs"
    aria-orientation="horizontal"
    id="horizontal-tabs-example"
    role="tablist"
  >
    <div
      class="emotion-3"
      display="flex"
    >
      <span
        aria-controls="horizontal-tabs-example-4"
        aria-selected="false"
        class="emotion-0"
        color="colorTextWeak"
        cursor="pointer"
        display="inline-block"
        font-size="fontSize30"
        font-weight="fontWeightSemibold"
        id="horizontal-tabs-example-1"
        overflow="hidden"
        role="tab"
        tabindex="-1"
      >
        Tab 1
      </span>
      <span
        aria-controls="horizontal-tabs-example-5"
        aria-disabled="true"
        aria-selected="false"
        class="emotion-1"
        color="colorTextWeak"
        cursor="not-allowed"
        display="inline-block"
        font-size="fontSize30"
        font-weight="fontWeightSemibold"
        id="horizontal-tabs-example-2"
        overflow="hidden"
        role="tab"
        tabindex="-1"
      >
        Tab 2
      </span>
      <span
        aria-controls="horizontal-tabs-example-6"
        aria-selected="true"
        class="emotion-0"
        color="colorTextWeak"
        cursor="pointer"
        display="inline-block"
        font-size="fontSize30"
        font-weight="fontWeightSemibold"
        id="1"
        overflow="hidden"
        role="tab"
        tabindex="0"
      >
        Tab 3
      </span>
    </div>
  </div>
  .emotion-0 {
  box-sizing: border-box;
  width: 100%;
}

<div
    class="emotion-0"
    width="100%"
  >
    <div
      aria-labelledby="horizontal-tabs-example-1"
      hidden=""
      id="horizontal-tabs-example-4"
      role="tabpanel"
      style="display: none;"
      tabindex="0"
    >
      Tab 1
    </div>
    <div
      aria-labelledby="horizontal-tabs-example-2"
      hidden=""
      id="horizontal-tabs-example-5"
      role="tabpanel"
      style="display: none;"
      tabindex="0"
    >
      Tab 2
    </div>
    <div
      aria-labelledby="1"
      id="horizontal-tabs-example-6"
      role="tabpanel"
      style=""
      tabindex="0"
    >
      Tab 3
    </div>
  </div>
  .emotion-0 {
  margin: 0;
  padding: 0;
  margin-bottom: space70;
  font-size: fontSize30;
  line-height: lineHeight40;
  font-weight: fontWeightNormal;
  color: colorText;
}

<p
    class="emotion-0"
    color="colorText"
    font-size="fontSize30"
    font-weight="fontWeightNormal"
  >
    This changes the keys to jump between tabs from left/right arrow to up/down arrow
  </p>
  .emotion-4 {
  box-sizing: border-box;
  width: 100%;
}

.emotion-5 {
  box-sizing: border-box;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: flex;
  min-width: size0;
}

.emotion-3 {
  box-sizing: border-box;
  border-left-width: borderWidth10;
  border-left-color: colorBorderWeak;
  border-left-style: solid;
  margin-right: space110;
  min-width: size20;
  max-width: size40;
}

.emotion-0 {
  box-sizing: border-box;
  border-left-color: transparent;
  border-left-style: solid;
  border-left-width: borderWidth20;
  color: colorTextWeak;
  display: block;
  margin-bottom: space40;
  padding-bottom: space30;
  padding-left: space50;
  padding-right: space50;
  padding-top: space30;
  cursor: pointer;
  font-size: fontSize30;
  font-weight: fontWeightSemibold;
  position: relative;
  -webkit-transition: border-color 100ms ease,color 100ms ease;
  transition: border-color 100ms ease,color 100ms ease;
}

.emotion-0:last-of-type {
  margin-bottom: space0;
}

.emotion-0[aria-selected=true] {
  border-left-color: colorBorderPrimary;
  color: colorTextLink;
}

.emotion-0:hover {
  border-left-color: colorBorderPrimaryStronger;
  color: colorTextLinkStronger;
}

.emotion-0:disabled,
.emotion-0:disabled:focus,
.emotion-0:disabled:hover,
.emotion-0[aria-disabled=true],
.emotion-0[aria-disabled=true]:focus,
.emotion-0[aria-disabled=true]:hover {
  border-left-color: transparent;
  color: colorTextWeaker;
}

.emotion-0:focus {
  border-left-color: colorBorderPrimaryStronger;
  box-shadow: shadowFocus;
  color: colorTextLinkStronger;
  outline: none;
}

<div
    class="emotion-5"
    display="flex"
  >
    <div
      aria-label="My tabs"
      aria-orientation="vertical"
      id="vertical-tabs-example"
      role="tablist"
    >
      <div
        class="emotion-3"
      >
        <span
          aria-controls="vertical-tabs-example-4"
          aria-selected="true"
          class="emotion-0"
          color="colorTextWeak"
          cursor="pointer"
          display="block"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
          id="2"
          role="tab"
          tabindex="0"
        >
          Tab 1 is a long tab name because the server sent a long tab name
        </span>
        <span
          aria-controls="vertical-tabs-example-5"
          aria-selected="false"
          class="emotion-0"
          color="colorTextWeak"
          cursor="pointer"
          display="block"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
          id="vertical-tabs-example-2"
          role="tab"
          tabindex="-1"
        >
          Tab 2
        </span>
        <span
          aria-controls="vertical-tabs-example-6"
          aria-selected="false"
          class="emotion-0"
          color="colorTextWeak"
          cursor="pointer"
          display="block"
          font-size="fontSize30"
          font-weight="fontWeightSemibold"
          id="vertical-tabs-example-3"
          role="tab"
          tabindex="-1"
        >
          Tab 3
        </span>
      </div>
    </div>
    <div
      class="emotion-4"
      width="100%"
    >
      <div
        aria-labelledby="2"
        id="vertical-tabs-example-4"
        role="tabpanel"
        style=""
        tabindex="0"
      >
        Tab 1
      </div>
      <div
        aria-labelledby="vertical-tabs-example-2"
        hidden=""
        id="vertical-tabs-example-5"
        role="tabpanel"
        style="display: none;"
        tabindex="0"
      >
        Tab 2
      </div>
      <div
        aria-labelledby="vertical-tabs-example-3"
        hidden=""
        id="vertical-tabs-example-6"
        role="tabpanel"
        style="display: none;"
        tabindex="0"
      >
        Tab 3
      </div>
    </div>
  </div>
</DocumentFragment>
`;
